Sources/StreamChat/WebSocketClient/Engine/URLSessionWebSocketEngine.swift

```diff
@@ -65,46 +79,75 @@ class URLSessionWebSocketEngine: NSObject, WebSocketEngine, URLSessionDataDelega
             }
         }
     }
+
+    private func makeURLSessionDelegateHandler() -> URLSessionDelegateHandler {
+        let urlSessionDelegateHandler = URLSessionDelegateHandler()
+        urlSessionDelegateHandler.didOpenWith = { [weak self] _ in
+            self?.callbackQueue.async {
+                self?.delegate?.webSocketDidConnect()
+            }
+        }
+
+        urlSessionDelegateHandler.didCloseWith = { [weak self] closeCode, reason in
+            var error: WebSocketEngineError?
+
+            if let reasonData = reason, let reasonString = String(data: reasonData, encoding: .utf8) {
+                error = WebSocketEngineError(
+                    reason: reasonString,
+                    code: closeCode.rawValue,
+                    engineError: nil
+                )
+            }
+
+            self?.callbackQueue.async { [weak self] in
+                self?.delegate?.webSocketDidDisconnect(error: error)
+            }
+        }
+
+        urlSessionDelegateHandler.didCompleteWith = { [weak self] error in
+            // If we received this callback because we closed the WS connection
+            // intentionally, `error` param will be `nil`.
+            // Delegate is already informed with `didCloseWith` callback,
+            // so we don't need to call delegate again.
+            guard let error = error else { return }
+
+            self?.callbackQueue.async { [weak self] in
+                self?.delegate?.webSocketDidDisconnect(error: WebSocketEngineError(error: error))
+            }
+        }
+
+        return urlSessionDelegateHandler
+    }
+
+    deinit {
+        disconnect()
+    }
+}
     
+@available(iOS 13, *)
+class URLSessionDelegateHandler: NSObject, URLSessionDataDelegate, URLSessionWebSocketDelegate {
+    var didOpenWith: ((_ protocol: String?) -> Void)?
+    var didCloseWith: ((_ code: URLSessionWebSocketTask.CloseCode, _ reason: Data?) -> Void)?
+    var didCompleteWith: ((Error?) -> Void)?
```

Comment:
Should we name them `onCompletion`/`onClose`/ etc? The current naming is imperative, while we are just creating blocks for later execution
We might need to standarize this naming convention, as we are using different patterns now.

Reply:
Yes exactly, makes total sense, will change this 👍